pkg/controller/tidbcluster/pod_control.go

```diff
@@ -0,0 +1,334 @@
+			return reconcile.Result{}, nil
+		}
+	}
+
+	if ok {
+		evictStatus := &v1alpha1.EvictLeaderStatus{
+			PodCreateTime: pod.CreationTimestamp,
+			Value:         value,
+		}
+		nowStatus := tc.Status.TiKV.EvictLeader[pod.Name]
+		if nowStatus == nil || *nowStatus != *evictStatus {
+			if tc.Status.TiKV.EvictLeader == nil {
+				tc.Status.TiKV.EvictLeader = make(map[string]*v1alpha1.EvictLeaderStatus)
+			}
+			tc.Status.TiKV.EvictLeader[pod.Name] = evictStatus
+			var err error
+			key := fmt.Sprintf("%s/%s", tc.Namespace, tc.Name)
+			tc, err = c.deps.Clientset.PingcapV1alpha1().TidbClusters(tc.Namespace).Update(ctx, tc, metav1.UpdateOptions{})
+			if err != nil {
+				return reconcile.Result{}, perrors.Annotatef(err, "failed to update tc %q status", key)
+			}
+
+			stat := c.getPodStat(pod)
+			stat.observeAnnotationCounts++
+			c.setPodStat(pod, stat)
+		}
+
+		pdClient := c.getPDClient(tc)
+		storeID, err := member.TiKVStoreIDFromStatus(tc, pod.Name)
+		if err != nil {
+			return reconcile.Result{}, perrors.Annotatef(err, "failed to get tikv store id from status for pod %s/%s", pod.Namespace, pod.Name)
+		}
+		err = pdClient.BeginEvictLeader(storeID)
+		if err != nil {
+			return reconcile.Result{}, perrors.Annotatef(err, "failed to evict leader for store %d (Pod %s/%s)", storeID, pod.Namespace, pod.Name)
+		}
+
+		if value == v1alpha1.EvictLeaderValueDeletePod {
+			tlsEnabled := tc.IsTLSClusterEnabled()
+			kvClient := c.deps.TiKVControl.GetTiKVPodClient(tc.Namespace, tc.Name, pod.Name, tlsEnabled)
+			leaderCount, err := kvClient.GetLeaderCount()
+			if err != nil {
+				return reconcile.Result{}, perrors.Annotatef(err, "failed to get leader count for pod %s/%s", pod.Namespace, pod.Name)
+			}
+
+			klog.Infof("Region leader count is %d for Pod %s/%s", leaderCount, pod.Namespace, pod.Name)
+
+			if leaderCount == 0 {
+				err = c.deps.KubeClientset.CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+				if err != nil && !errors.IsNotFound(err) {
+					return reconcile.Result{}, perrors.Annotatef(err, "failed to delete pod %q", pod.Name)
+				}
+			} else {
+				// re-check leader count next time
+				return reconcile.Result{RequeueAfter: c.recheckLeaderCountDuration}, nil
+			}
+		}
+	} else {
+		// 1. delete evict-leader scheduler
+		// 2. delete pod from tc.Status.TiKV.EvictLeader and update it to api-server
+		endEvict := func() error {
+			pdClient := c.getPDClient(tc)
+			storeID, err := member.TiKVStoreIDFromStatus(tc, pod.Name)
+			if err != nil {
+				return perrors.Annotatef(err, "failed to get tikv store id from status for pod %s/%s", pod.Namespace, pod.Name)
+			}
+
+			err = pdClient.EndEvictLeader(storeID)
+			if err != nil {
+				return perrors.Annotatef(err, "failed to remove evict leader scheduler for store %d, pod %s/%s", storeID, pod.Namespace, pod.Name)
+			}
+
+			err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
The logic should be:
1. update TC
2. if any error, get the TC, update the status for this Pod

If it's decided to not retry in L234, it's ok not to retry here.